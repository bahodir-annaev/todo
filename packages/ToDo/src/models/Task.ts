import { Record } from 'immutable';
import { Priorities } from '../constants';

interface ITaskParams {
  complete?: boolean;
  description?: string;
  id?: number;
  priority?: Priorities;
}

let id = 0;
export class Task extends Record({
  complete: false,
  description: '',
  id: 0,
  priority: Priorities.PRIORITY_NORMAL,
}) {
  constructor(props: ITaskParams) {
    super(props);
  }

  static create(props: ITaskParams): Task {
    const task = props.id ? new Task({ ...props }) : new Task({ ...props, id: id++ });

    return task;
  }

  toggle() {
    return this.set('complete', !this.complete) as this;
  }
}
