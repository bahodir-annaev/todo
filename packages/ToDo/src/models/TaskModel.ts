import { Record } from 'immutable';
import { Priorities } from '../constants';

let id = 0;
export class TaskModel extends Record({
  description: '',
  finished: false,
  id: 0,
  priority: Priorities.PRIORITY_NORMAL,
}) {
  constructor(props: Partial<TaskModel>) {
    super(props);

    if (props && !props.id) {
      return this.set('id', id++);
    }
  }

  static create(props: Partial<TaskModel>) {
    return new TaskModel(props);
  }

  toggle() {
    return this.set('finished', !this.finished);
  }
}
