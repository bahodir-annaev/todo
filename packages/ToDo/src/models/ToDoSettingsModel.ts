import { Record } from 'immutable';

export const enum applyColorTo {
  background = 'background',
  text = 'text',
}

export class TaskAppearanceRecord extends Record({
  applyTo: applyColorTo.background,
  color: 'grey',
}) {}

export class TodoAppearanceRecord extends Record({ finishedTask: new TaskAppearanceRecord() }) {}

export class TodoFunctionalityRecord extends Record({ filtering: true, priority: true }) {}

const DEFAULT_APPEARANCE = new TodoAppearanceRecord();
const DEFAULT_FUNCTIONALITY = new TodoFunctionalityRecord();

export class TodoSettingsModel extends Record({
  appearance: DEFAULT_APPEARANCE,
  functionality: DEFAULT_FUNCTIONALITY,
}) {
  static create(settings?: ITodoSettingsModel) {
    if (settings) {
      const appearance =
        settings.appearance && settings.appearance.finishedTask
          ? new TodoAppearanceRecord({
              finishedTask: new TaskAppearanceRecord(settings.appearance.finishedTask),
            })
          : DEFAULT_APPEARANCE;

      const functionality = settings.functionality
        ? new TodoFunctionalityRecord(settings.functionality)
        : DEFAULT_FUNCTIONALITY;

      return new TodoSettingsModel({ appearance, functionality });
    } else {
      return new TodoSettingsModel();
    }
  }
}

interface ITodoAppearance {
  finishedTask: TaskAppearanceRecord;
}

interface ITodoFunctionality {
  filtering: boolean;
  priority: boolean;
}

interface ITodoSettingsModel {
  appearance?: ITodoAppearance;
  functionality?: ITodoFunctionality;
}
