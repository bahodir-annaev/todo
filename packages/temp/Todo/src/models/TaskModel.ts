import { Record } from 'immutable';
import { Priorities } from '../constants';

let id = 0;
export class TaskModel extends Record({
  description: '',
  finished: false,
  id: 0,
  priority: Priorities.PRIORITY_NORMAL,
}) {
  constructor(props?: Partial<TaskModel>) {
    super(props);

    return props && !props.id ? this.set('id', id++) : this;
  }

  static create(props: Partial<TaskModel>) {
    return new TaskModel(props);
  }

  toggle() {
    return this.set('finished', !this.finished);
  }
}
