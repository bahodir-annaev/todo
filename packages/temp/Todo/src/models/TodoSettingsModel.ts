import { Record } from 'immutable';

export const enum applyColorTo {
  background = 'background',
  text = 'text',
}

export class TaskAppearance extends Record({
  applyTo: applyColorTo.background,
  color: 'grey',
}) {}

export class TodoAppearance extends Record({ finishedTask: new TaskAppearance() }) {}

export class TodoFunctionality extends Record({ filtering: true, priority: true }) {}

const DEFAULT_APPEARANCE = new TodoAppearance();
const DEFAULT_FUNCTIONALITY = new TodoFunctionality();

export class TodoSettingsModel extends Record({
  appearance: DEFAULT_APPEARANCE,
  functionality: DEFAULT_FUNCTIONALITY,
}) {
  static create(settings?: ITodoSettingsModel) {
    if (settings) {
      const appearance =
        settings.appearance && settings.appearance.finishedTask
          ? new TodoAppearance({
              finishedTask: new TaskAppearance(settings.appearance.finishedTask),
            })
          : DEFAULT_APPEARANCE;

      const functionality = settings.functionality
        ? new TodoFunctionality(settings.functionality)
        : DEFAULT_FUNCTIONALITY;

      return new TodoSettingsModel({ appearance, functionality });
    } else {
      return new TodoSettingsModel();
    }
  }
}

interface ITodoAppearance {
  finishedTask?: TaskAppearance;
}

interface ITodoFunctionality {
  filtering?: boolean;
  priority?: boolean;
}

interface ITodoSettingsModel {
  appearance?: ITodoAppearance;
  functionality?: ITodoFunctionality;
}
