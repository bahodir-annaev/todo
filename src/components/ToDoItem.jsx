import * as React from "react";

export class ToDoItem extends React.Component {
    render(){
        return (
            <div>
                <input type="checkbox" onClick={this.props.toggleComplete}/>&nbsp;
                <span className={this.props.task.complete ? "complete": ""}> {this.props.task.description}</span>&nbsp;
                <input type="button" onClick={this.props.removeTask}  value="Remove"/>
            </div>
        );
    }
}