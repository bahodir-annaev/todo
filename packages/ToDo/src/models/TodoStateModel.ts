import { OrderedMap, Record } from 'immutable';
import { Filters } from '../constants';
import { TaskModel } from './TaskModel';

export class TodoStateModel extends Record({
  activeFilter: Filters.ALL,
  tasks: OrderedMap<number, TaskModel>(),
}) {}
