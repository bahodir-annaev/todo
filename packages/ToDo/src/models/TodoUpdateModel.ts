import { Filters } from '../constants';
import { TaskModel } from './TaskModel';

export interface ITodoUpdateModel {
  data: number | Filters | TaskModel;
  type: string;
}
