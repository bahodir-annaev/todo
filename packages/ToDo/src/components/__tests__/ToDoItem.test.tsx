import { shallow } from 'enzyme';
import * as React from 'react';

import { Task } from '../../models/Task';
import { applyColorTo } from '../../models/ToDoSettingsModel';
import { ToDoItem } from '../ToDoItem';

const setup = (propsToChange: object, taskProperties?: object) => {
  const applyTo = applyColorTo.text;
  const props = Object.assign(
    {
      appearance: {
        finishedTask: {
          applyTo,
          color: 'grey',
        },
      },
      removeTask: jest.fn(),
      task: Object.assign({ ...Task.create({ description: 'Test Task' }) }, taskProperties),
      toggleComplete: jest.fn(),
    },
    propsToChange,
  );

  const todoItem = shallow(<ToDoItem {...props} />);

  return {
    props,
    todoItem,
  };
};

describe('ToDoItem component', () => {
  test('should render', () => {
    const { todoItem } = setup({});
    expect(todoItem).toMatchSnapshot();
  });

  test('toggleComplete method should be called', () => {
    const { props, todoItem } = setup({});
    todoItem.find('input').simulate('change');

    expect(props.toggleComplete).toHaveBeenCalled();
  });

  test('removeTask method should be called', () => {
    const { props, todoItem } = setup({});
    todoItem.find('button').simulate('click');

    expect(props.removeTask).toHaveBeenCalled();
  });

  test('completed task should be checked and should have finished className', () => {
    const { todoItem } = setup({}, { complete: true });
    expect(todoItem.find('input').props().checked).toBe(true);
    expect(todoItem.find('span').hasClass('complete')).toBe(true);
  });
});
