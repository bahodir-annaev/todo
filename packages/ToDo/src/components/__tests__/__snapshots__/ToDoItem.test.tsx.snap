// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component should render 1`] = `
<div
  className="item-container"
>
  <input
    checked={false}
    onChange={[MockFunction]}
    type="checkbox"
  />
  <span
    className=" item-desc"
    style={Object {}}
  >
    Test Task
  </span>
  <button
    onClick={[MockFunction]}
  >
    Remove
  </button>
</div>
`;
